usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1372,6 +1435,34 @@ zil_lwb_write_issue(zilog_t *zilog, lwb_t *lwb)
 	lwb->lwb_issued_timestamp = gethrtime();
 	lwb->lwb_state = LWB_STATE_ISSUED;
 
+	/*
+	 * If this lwb does not have any threads waiting for it to
+	 * complete, we want to defer issuing the DKIOCFLUSHWRITECACHE
+	 * command to the vdevs written to by "this" lwb, and instead
+	 * rely on the "next" lwb to handle the DKIOCFLUSHWRITECACHE
+	 * command for those vdevs. Thus, we merge the vdev tree of
+	 * "this" lwb with the vdev tree of the "next" lwb in the list,
+	 * and assume the "next" lwb will handle flushing the vdevs (or
+	 * deferring the flush(s) again).
+	 *
+	 * This is a necessary performance optimization, especially for
+	 * workloads with lots of async write activity and few sync
+	 * write and/or fsync activity, as it has the potential to
+	 * coalesce multiple flush commands to a vdev into one.
+	 */
+	boolean_t defer_flush =
+	    (list_head(&lwb->lwb_waiters) == NULL && nlwb != NULL);
+
+	if (defer_flush) {
+		ASSERT3S(nlwb->lwb_state, ==, LWB_STATE_CLOSED);
+		while ((zv = avl_destroy_nodes(t, &cookie)) != NULL) {
```

Comment:
Should `lwb->lwb_vdev_lock` be held here?  I think a `zfs_get_done()` callback could possibly race with us here?

Reply:
We grab the lock inside of the `zil_lwb_add_vdev()` function.

With that said, I think if another thread could concurrently add vdevs to the tree, then we have a problem in the design, since those vdevs added by `zfs_get_done()` would not get deferred. Any vdevs added by `zfs_get_done()` must be added prior to executing this logic in order for those to get deferred (I think the vdevs would still get flushed, they just might not be flushed in the most optimal way).

Reply:
So, taking a few steps back here...

With `nlwb` assured to be in the `CLOSED` state, and `zl_issuer_lock` held anyways, we needn't worry about grabbing `nlwb->lwb_vdev_lock`, since there should be no threads capable of adding to it yet.  On the other hand, any `WR_INDIRECT` writes to the current `lwb` do have the possibility of adding vdevs to its tree.  The short-and-quick fix would be to simply grab `lwb->lwb_vdev_lock` which would prevent a nasty race on the AVL, but would result in high flush rates for any dataset undergoing a lot of `WR_INDIRECT` IO.  Alternatively, we could reexamine how those indirect blocks are being designated for flush on the `lwb` to begin with.

Rather than making the `zil_lwb_add_vdev()` call for the `blkptr_t` instances when the write completes, as is done now for `WR_INDIRECT` operations, we could change the logic in `dmu_sync()` (and `dmu_sync_late_arrival()`) to do the vdev association prior to issuing the write.  Doing so would eliminate the need for `lwb_vdev_lock` entirely, since manipulation of the vdev tree would be done entirely under `zl_issuer_lock` (Or later under `zl_lock` when cleaning up the tree during flush or completion).

Thoughts?